baseplate/secrets/store.py

```diff
@@ -0,0 +1,253 @@
+
+
+class CorruptSecretError(Exception):
+    """Raised when the requested secret does not match the expected format."""
+
+    def __init__(self, path, message):
+        super(CorruptSecretError, self).__init__()
+        self.path = path
+        self.message = message
+
+    def __str__(self):  # pragma: nocover
+        return "{}: {}".format(self.path, self.message)
+
+
+_VersionedSecret = collections.namedtuple(
+    "VersionedSecret", "previous current next")
+
+
+class VersionedSecret(_VersionedSecret):
+    """A versioned secret.
+
+    Versioned secrets allow for seamless rotation of keys. When using the
+    secret to generate tokens (e.g. signing a message) always use the
+    ``current`` value. When validating tokens, check against all the versions
+    in ``all_versions``. This will allow keys to rotate smoothly even if not
+    done instantly across all users of the secret.
+
+    """
+
+    @property
+    def all_versions(self):
+        """Return an iterator over the available versions of this secret."""
+        yield self.current
+
+        if self.previous is not None:
+            yield self.previous
+
+        if self.next is not None:
+            yield self.next
+
+    @classmethod
+    def from_simple_secret(cls, value):
+        """Make a fake versioned secret from a single value.
+
+        This is a backwards compatibility shim for use with APIs that take
+        versioned secrets. Try to use proper versioned secrets fetched from the
+        secrets store instead.
+
+        """
+        return cls(
+            previous=None,
+            current=value,
+            next=None,
+        )
+
+
+def _decode_secret(path, encoding, value):
+    if encoding == "identity":
+        # encode to bytes for consistency with the base64 path. utf-8 because
+        # that undoes json encoding.
+        return value.encode("utf-8")
+    elif encoding == "base64":
+        try:
+            return base64.b64decode(value)
+        except (TypeError, binascii.Error) as exc:
+            raise CorruptSecretError(path, "unable to decode base64: %s" % exc)
+    else:
+        raise CorruptSecretError(path, "unknown encoding: %r" % encoding)
+
+
+class SecretsStore(ContextFactory):
+    """Access to secret tokens with automatic refresh on expiration.
+
+    This local vault allows access to the secrets cached on disk by the fetcher
+    daemon. It will automatically reload the cache when it expires. Do not
+    cache or store the values returned by this class's methods but rather get
+    them from this class each time you need them. The secrets are served from
+    memory so there's little performance impact to doing so and you will be
+    sure to always have the current version in the face of key rotation etc.
+
+    """
+
+    def __init__(self, path):
+        self._path = path
+        self._expiration = datetime.datetime.min
+        self._secrets = None
+        self._vault_token = None
+
+    def _load_if_expired(self):
```

Comment:
Should we force a reload if the mtime is higher than when we last read the file even if not expired yet? That would mean extra `stat` calls, but that should be in-cache most of the time. The upside would be that we are sure we get the latest copy if someone restarts the fetcher (e.g. for new configs). Alternatively, we punt and say "if you really need new data, send an app restart".